README.md
![Xiaomi Mijia LYWSD03MMC](https://i.postimg.cc/C5VYY82F/gatttool-mijia.jpg)

# Xiaomi Mijia LYWSD03MMC

[![Build Status](https://img.shields.io/travis/com/r-hede/xiaomi-mijia-lywsd03mmc?style=flat-square)](https://app.travis-ci.com/r-hede/xiaomi-mijia-lywsd03mmc)
[![Coverage Status](https://img.shields.io/coveralls/github/r-hede/xiaomi-mijia-lywsd03mmc?style=flat-square)](https://coveralls.io/github/r-hede/xiaomi-mijia-lywsd03mmc)
[![Code style](https://img.shields.io/badge/code_style-standard-397e96.svg?style=flat-square)](https://standardjs.com)
[![Version](https://img.shields.io/github/package-json/v/r-hede/xiaomi-mijia-lywsd03mmc?color=397e96&style=flat-square)](https://github.com/r-hede/xiaomi-mijia-lywsd03mmc/releases/latest)
[![License](https://img.shields.io/github/license/r-hede/xiaomi-mijia-lywsd03mmc?color=397e96&style=flat-square)](LICENSE)

Search and get data from nearby sensors.

Only works with the Xiaomi Mijia Bluetooth temperature and humidity sensor (LYWSD03MMC).

Developed and tested with love on Raspberry Pi 4 and Raspberry Pi Zero W.

The sudo command is not required.

## Installation

```
npm i xiaomi-mijia-lywsd03mmc
```

## Quick Start Example

```js
const lywsd03mmc = require('xiaomi-mijia-lywsd03mmc')

// Get an array with the MAC addresses of the nearby sensors
lywsd03mmc.getSensors().then((res) => {
    res.forEach(sensorAddress => {
        // Get an object with the sensor data
        lywsd03mmc.getData(sensorAddress).then((res) => {
            // {
            //     address: 'XX:XX:XX:XX:XX:XX',
            //     humidityLevel: XX,
            //     temperature: XX,
            //     batteryLevel: XX
            // }
            console.log(res)
        }).catch((err) => {
            console.error(`[xiaomi-mijia-lywsd03mmc] Unable to get data (address: ${sensorAddress}, error: ${err})`)
        })
    })
}).catch((err) => {
    console.error(`[xiaomi-mijia-lywsd03mmc] Unable to get sensors (error: ${err})`)
})
```

## Documentation

<dl>
<dt><a href="#getSensors">getSensors(timeoutDuration)</a> ⇒ <code>Promise.&lt;Array.&lt;string&gt;&gt;</code></dt>
<dd><p>Scan and get the MAC addresses of the nearby sensors.</p>
</dd>
<dt><a href="#getData">getData(sensorAddress, tempInFahrenheit, timeoutDuration)</a> ⇒ <code>Promise.&lt;Object&gt;</code></dt>
<dd><p>Get the data from the specified sensor (temperature, humidity level, battery level) every 20 seconds during 2 minutes. Stops when data are got.</p>
</dd>
</dl>

<a name="getSensors"></a>

## getSensors(timeoutDuration) ⇒ <code>Promise.&lt;Array.&lt;string&gt;&gt;</code>
Scan and get the MAC addresses of the nearby sensors.

**Kind**: global function
**Returns**: <code>Promise.&lt;Array.&lt;string&gt;&gt;</code> - Promise containing an array with the MAC addresses of the detected sensors.

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| timeoutDuration | <code>number</code> | <code>10</code> | Maximum time in seconds assigned to the scan |

**Example**
```js
getSensors().then((res) => {
    console.log(res) // ['XX:XX:XX:XX:XX:XX', 'XX:XX:XX:XX:XX:XX', ..]
}).catch((err) => {
    console.error(`Unable to get sensors (error: ${err})`)
})
```
<a name="getData"></a>

## getData(sensorAddress, tempInFahrenheit, timeoutDuration) ⇒ <code>Promise.&lt;Object&gt;</code>
Get the data from the specified sensor (temperature, humidity level, battery level) every 20 seconds during 2 minutes. Stops when data are got.

**Kind**: global function
**Returns**: <code>Promise.&lt;Object&gt;</code> - Promise containing an object with the requested data.

| Param | Type | Default | Description |
| --- | --- | --- | --- |
| sensorAddress | <code>string</code> |  | MAC address of the sensor to connect |
| tempInFahrenheit | <code>boolean</code> | <code>false</code> | Temperature output format (Celsius/Fahrenheit) |
| timeoutDuration | <code>number</code> | <code>2</code> | Maximum time in minutes assigned to the connection |

**Example**
```js
// Replace with a valid MAC address of a nearby sensor
var sensorAddress = 'XX:XX:XX:XX:XX:XX'
getData(sensorAddress).then((res) => {
    // {
    //     address: 'XX:XX:XX:XX:XX:XX',
    //     humidityLevel: XX,
    //     temperature: XX,
    //     batteryLevel: XX
    // }
    console.log(res)
}).catch(() => {
    console.error(`[xiaomi-mijia-lywsd03mmc] Unable to get data (address: ${sensorAddress}, error: ${err})`)
})
```

## Change Log

This project adheres to [Semantic Versioning](https://semver.org).

Every release is documented on the GitHub [Releases](https://github.com/r-hede/xiaomi-mijia-lywsd03mmc/releases) page.

## License

[MIT](LICENSE)

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2021-11-05
### Changed
- Replacement of the "coveralls" library by the "coveralls-next" library.
- Updated versions of Node.js we can work with.
- Minor improvements (addition of tests, respect of naming conventions...). 

## [1.0.0] - 2021-10-15
### Added
- First public version of the project.
- That’s one small step for a man, one giant leap for me.

index.test.js
const lywsd03mmc = require('./index')

describe('getSensors', () => {
  it('should return an array of string, an empty array or a string', () => {
    return lywsd03mmc.getSensors().then((res) => {
      expect(typeof res[0] === 'string' || expect(res).toEqual(expect.any(Array))).toBeTruthy()
    }).catch((err) => {
      expect(err).toEqual(
        expect.any(String)
      )
    })
  }, 15000)
})

describe('getData', () => {
  it('should return an object or a string', () => {
    return lywsd03mmc.getData('XX:XX:XX:XX:XX:XX').then((res) => {
      expect(res).toEqual(
        expect.any(Object)
      )
    }).catch((err) => {
      expect(err).toEqual(
        expect.any(String)
      )
    })
  }, 140000)
})

describe('_getHumidityLevel', () => {
  it('should return the number 68', () => {
    const buf = Buffer.from('460844c00a', 'hex')
    return expect(lywsd03mmc._getHumidityLevel(buf)).toBe(68)
  })
})

describe('_getTemperature', () => {
  it('should return the number 21.2', () => {
    const buf = Buffer.from('460844c00a', 'hex')
    return expect(lywsd03mmc._getTemperature(buf)).toBe(21.2)
  })
  it('should return the number -5.2', () => {
    const buf = Buffer.from('ccff44c00a', 'hex')
    return expect(lywsd03mmc._getTemperature(buf)).toBe(-5.2)
  })
})

describe('_getBatteryLevel', () => {
  it('should return the number 72', () => {
    const buf = Buffer.from('460844c00a', 'hex')
    return expect(lywsd03mmc._getBatteryLevel(buf)).toBe(72)
  })
})

describe('_map', () => {
  it('should return the number 44.4', () => {
    return expect(lywsd03mmc._map(2500, 2100, 3000, 0, 100)).toBe(44.4)
  })
  it('should return the number 100', () => {
    return expect(lywsd03mmc._map(4000, 2100, 3000, 0, 100)).toBe(100)
  })
  it('should return the number 0', () => {
    return expect(lywsd03mmc._map(2000, 2100, 3000, 0, 100)).toBe(0)
  })
})


index.js
const { exec } = require('child_process')

module.exports = (() => {
  /**
   * Scan and get the MAC addresses of the nearby sensors.
   *
   * @param {number} timeoutDuration=10 - Maximum time in seconds assigned to the scan.
   *
   * @example
   * getSensors().then((res) => {
   *     console.log(res) // ['XX:XX:XX:XX:XX:XX', 'XX:XX:XX:XX:XX:XX', ..]
   * }).catch((err) => {
   *     console.error(`Unable to get sensors (error: ${err})`)
   * })
   *
   * @returns {Promise<Array<string>>} - Promise containing an array with the MAC addresses of the detected sensors.
   */
  function getSensors (timeoutDuration = 10) {
    return new Promise((resolve, reject) => {
      const sensorsAddresses = []
      const query = 'echo $(sudo ' +
        'timeout -s SIGINT ' +
        timeoutDuration +
        ' hcitool lescan)'
      exec(query, (err, stdout, stderr) => {
        if (err) {
          reject(err.message)
        } else if (stderr) {
          reject(stderr.slice(0, -1))
        } else {
          // Remove the useless beginning of the string and the last line break
          stdout = stdout.slice(0, -1).substring(12)
          // Split the string into an array of substrings to browse the results
          stdout = stdout.split(' ')
          for (let i = 0; i < stdout.length; i++) {
            if (stdout[i] === 'LYWSD03MMC') sensorsAddresses.push(stdout[i - 1])
          }
          resolve(sensorsAddresses)
        }
      })
    })
  }

  /**
   * Get the data from the specified sensor (temperature, humidity level,
   * battery level) every 20 seconds during 2 minutes. Stops when data are got.
   *
   * @param {string} sensorAddress - MAC address of the sensor to connect.
   * @param {boolean} tempInFahrenheit=false - Temperature output format (Celsius/Fahrenheit).
   * @param {number} timeoutDuration=2 - Maximum time in minutes assigned to the connection.
   *
   * @example
   * // Replace with a valid MAC address of a nearby sensor
   * const sensorAddress = 'XX:XX:XX:XX:XX:XX'
   * getData(sensorAddress).then((res) => {
   *     // {
   *     //     address: 'XX:XX:XX:XX:XX:XX',
   *     //     humidityLevel: XX,
   *     //     temperature: XX,
   *     //     batteryLevel: XX
   *     // }
   *     console.log(res)
   * }).catch(() => {
   *     console.error(`[xiaomi-mijia-lywsd03mmc] Unable to get data (address: ${sensorAddress}, error: ${err})`)
   * })
   *
   * @returns {Promise<Object>} - Promise containing an object with the requested data.
   */
  function getData (sensorAddress, tempInFahrenheit = false, timeoutDuration = 2) {
    return new Promise((resolve, reject) => {
      // Try to get the data every 20 seconds during {timeoutDuration} minute(s)
      for (let timeouts = [], i = 0; i <= timeoutDuration * 3; i++) {
        (function (index) {
          // Keep all timers to delete them later
          timeouts.push(setTimeout(() => {
            // Get the temperature, the humidity level and the battery level
            _listenHandle(sensorAddress).then((res) => {
              // Delete all existing timers
              for (let j = 0; j < timeouts.length; j++) {
                clearTimeout(timeouts[j])
              }
              resolve({
                address: sensorAddress,
                humidityLevel: _getHumidityLevel(res),
                temperature: _getTemperature(res, tempInFahrenheit),
                batteryLevel: _getBatteryLevel(res)
              })
            }).catch((err) => {
              // No data received within the allocated time
              if (index === timeoutDuration * 3) {
                reject(err)
              } else {
                console.error(`[xiaomi-mijia-lywsd03mmc] ${err}`)
              }
            })
          }, i * 20000, i))
        })(i)
      }
    })
  }

  /**
   * Listen a handle for a few seconds in order to get the desired
   * data (temperature, humidity level, battery level).
   *
   * @param {string} sensorAddress - MAC address of the sensor to search for.
   *
   * @example
   * // Replace with a valid MAC address of a nearby sensor
   * const sensorAddress = 'XX:XX:XX:XX:XX:XX'
   * // Set to true if you want the temperature in Fahrenheit
   * const tempInFahrenheit = false
   * _listenHandle(sensorAddress).then((res) => {
   *     console.log({
   *         'humidityLevel': _getHumidityLevel(res),
   *         'temperature': _getTemperature(res, tempInFahrenheit),
   *         'batteryLevel': _getBatteryLevel(res)
   *     })
   * }).catch((err) => {
   *     console.error(err)
   * })
   *
   * @returns {Promise<string|Buffer>} - Promise with an error (string) or the requested data (Buffer).
   */
  function _listenHandle (sensorAddress) {
    return new Promise((resolve, reject) => {
      const query = 'timeout 15 gatttool -b ' +
        sensorAddress +
        ' --char-read' +
        ' -a 0x38' +
        ' --listen' +
        ' | grep --max-count=1 \'Notification handle\''
      exec(query, (err, stdout, stderr) => {
        if (err) {
          reject(err.message.slice(0, -1))
        } else if (stderr) {
          reject(stderr.slice(0, -1))
        } else {
          // Keep only the value and remove the spaces
          stdout = stdout.substr(36, 14).replace(/\s/g, '')
          // Set the data in a buffer
          stdout = Buffer.from(stdout, 'hex')
          resolve(stdout)
        }
      })
    })
  }

  /**
   * Read only the useful part of a buffer and convert the
   * hexadecimal value to a decimal value of the humidity level.
   *
   * @param {Buffer} buf - Buffer that contains the humidity level in hexadecimal value.
   *
   * @example
   * const buf = Buffer.from('460844c00a', 'hex')
   * const humidityLevel = _getHumidityLevel(buf)
   * console.log(humidityLevel) // 68
   *
   * @returns {number} - Humidity level in decimal value.
   */
  function _getHumidityLevel (buf) {
    return buf.readUInt8(2)
  }

  /**
   * Read only the useful part of a buffer and convert the
   * hexadecimal value to a decimal value of the temperature.
   *
   * @param {Buffer} buf - Buffer that contains the temperature in hexadecimal value.
   * @param {boolean} valueInFahrenheit=false - Temperature output format (Celsius/Fahrenheit).
   * @param {undefined} val - Internal variable of the function.
   *
   * @example
   * const buf = Buffer.from('460844c00a', 'hex')
   * const temperature = _getTemperature(buf)
   * console.log(temperature) // 21.2
   *
   * @returns {number} - Temperature in decimal value.
   */
  function _getTemperature (buf, valueInFahrenheit = false, val) {
    if (buf[1] === 255) {
      // Temperature is negative
      val = Number((-65536 + buf.readUInt16LE(0)) / 10)
    } else {
      // Temperature is positive
      val = Number((buf.readUInt16LE(0) / 100).toFixed(1))
    }
    return valueInFahrenheit ? Number((val * 1.8 + 32).toFixed(1)) : val
  }

  /**
   * Read only the useful part of a buffer and convert the
   * hexadecimal value to a decimal value of the battery level.
   *
   * @param {Buffer} buf - Buffer that contains the battery level in hexadecimal value.
   *
   * @example
   * const buf = Buffer.from('460844c00a', 'hex')
   * const batteryLevel = _getBatteryLevel(buf)
   * console.log(batteryLevel) // 72
   *
   * @returns {number} - Battery level in decimal value.
   */
  function _getBatteryLevel (buf) {
    return Math.round(_map(buf.readUInt16LE(3), 2100, 3000, 0, 100))
  }

  /**
   * Re-maps a number from one range to another.
   *
   * @param {number} val - Number to map.
   * @param {number} inMin - Lower bound of the value’s current range.
   * @param {number} inMax - Upper bound of the value’s current range.
   * @param {number} outMin - Lower bound of the value’s target range.
   * @param {number} outMax - Upper bound of the value’s target range.
   *
   * @example
   * const mappedValue = _map(2500, 2100, 3000, 0, 100)
   * console.log(mappedValue) // 44.4
   *
   * @returns {number} - Mapped value.
   */
  function _map (val, inMin, inMax, outMin, outMax) {
    if (val > inMax) {
      return outMax
    } else if (val < inMin) {
      return outMin
    } else {
      return Number(((val - inMin) * (outMax - outMin) / (inMax - inMin) + outMin).toFixed(1))
    }
  }

  return {
    getSensors: getSensors,
    getData: getData,
    _listenHandle: _listenHandle,
    _getHumidityLevel: _getHumidityLevel,
    _getTemperature: _getTemperature,
    _getBatteryLevel: _getBatteryLevel,
    _map: _map
  }
})()
